### PR TITLE
Clean up join test

### PR DESCRIPTION
**Pull request by mkleen**
This makes join test simpler and cleaner. The join logic moves now into the optimizer and this setup makes it easier to run the tests.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/planner/operators/JoinTest.java [hunk 1]
```diff
@@ -309,16 +324,18 @@ public void testHashJoin_TablesSwitchWhenRightBiggerThanLeft() {
...
+                                   "      ├ Collect[doc.locations | [id] | true]\n" +
+                                   "      └ Collect[doc.users | [name, id] | true]");
+
+        LogicalPlan hashjoin = plan.sources().get(0).sources().get(0);
         assertThat(hashjoin).isExactlyInstanceOf(HashJoin.class);
         assertThat(((HashJoin) hashjoin).lhs().getRelationNames())
             .as("Smaller table must be on the left-hand-side")
             .containsExactly(TEST_DOC_LOCATIONS_TABLE_IDENT);
-
-        Join join = buildJoin(hashjoin);
```
Comment (matriv) [inline server/src/test/java/io/crate/planner/operators/JoinTest.java:320]: Why did you remove this? not revalant?

File: server/src/test/java/io/crate/planner/operators/JoinTest.java [hunk 2]
```diff
@@ -288,14 +295,22 @@ public void testHashJoin_TableOrderInLogicalAndExecutionPlan() {
...
+            "    ├ Collect[doc.users | [name, id] | true]\n" +
+            "    └ Collect[doc.locations | [id] | true]"
+        );
+        var hashJoin = plan.sources().get(0);
+        assertThat(hashJoin).isExactlyInstanceOf(HashJoin.class);
+        assertThat(((HashJoin) hashJoin).rhs().getRelationNames())
             .as("Smaller table must be on the right-hand-side")
             .containsExactly(TEST_DOC_LOCATIONS_TABLE_IDENT);
 
-        Join join = buildJoin(operator);
```
Comment (matriv) [inline server/src/test/java/io/crate/planner/operators/JoinTest.java:297]: also this? I don't even get where this `other_id` comes from :)

**Review by matriv — COMMENTED** [review]
Thx! looks much nicer now!
Left a couple of questions/confusion for me.

File: server/src/test/java/io/crate/planner/operators/JoinTest.java (hunk 1 above)
Comment (mkleen) [reply server/src/test/java/io/crate/planner/operators/JoinTest.java:320]: The table swap used to be in the build method of hashjoin, but since now everything happens on logical plan level already there is no need anymore to check for the execution plan.

Comment (mkleen) [reply server/src/test/java/io/crate/planner/operators/JoinTest.java:320]: i could probably even remove more code.

Comment (mkleen) [reply server/src/test/java/io/crate/planner/operators/JoinTest.java:320]: i think for most cases the logical plan comparision would be enough.

**Comment by mkleen** [comment]
I will do the same for nl-joins once the logic moved.

**Review by matriv — APPROVED** [review]
thx!

File: server/src/test/java/io/crate/planner/operators/JoinTest.java [hunk 3]
```diff
@@ -288,14 +289,13 @@ public void testHashJoin_TableOrderInLogicalAndExecutionPlan() {
         rowCountByTable.put(TEST_DOC_LOCATIONS_TABLE_IDENT, new Stats(10, 0, Map.of()));
         e.updateTableStats(rowCountByTable);
 
-        LogicalPlan operator = createLogicalPlan(mss);
-        assertThat(operator).isExactlyInstanceOf(HashJoin.class);
-        assertThat(((HashJoin) operator).rhs().getRelationNames())
-            .as("Smaller table must be on the right-hand-side")
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/planner/operators/JoinTest.java:294]: Would be cool if we can preserve things like this `.as` in the new assertion. Afaik it's still accurate and useful information that clarifies the test intention?

Same in another test case further below

Comment (mkleen) [reply server/src/test/java/io/crate/planner/operators/JoinTest.java:294]: ok, will do.

**Comment by mkleen** [comment]
retest this please
